proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+
+## Stakeholders and Reviewers
+
+- [@dotnet/interop-contrib](https://github.com/orgs/dotnet/teams/interop-contrib)
+- [@dotnet/macios](https://github.com/orgs/dotnet/teams/macios)
+- [@lambdageek](https://github.com/lambdageek)
+- [@SamMonoRT](https://github.com/SamMonoRT)
+- [@kotlarmilos](https://github.com/kotlarmilos)
+- [@JulieLeeMSFT](https://github.com/JulieLeeMSFT)
+- [@amanasifkhalid](https://github.com/amanasifkhalid)
+- [@davidortinau](https://github.com/davidortinau)
+- [@jkotas](https://github.com/jkotas)
+
+## Design
+
+### Runtime
+
+#### Calling Conventions
+
+Swift has two calling conventions at its core, the "Swift" calling convention, and the "SwiftAsync" calling convention. We will begin by focusing on the "Swift" calling convention, as it is the most common. The Swift calling convention has a few elements that we need to contend with. Primarily, it allows passing arguments in more registers. Additionally, it has two dedicated registers for specific components of the calling convention, the self register and the error register. The runtime support for this calling convention must support all of these features. The additional registers per argument is relatively easy to support, as each of our compilers must already have some support for this concept to run on Linux or Apple platforms today. We have a few options for the remaining two features we need to support.
+
+The "SwiftAsync" calling convention has an additional "async context" register and specific behaviors expected around how the Swift async state machine is called. Full investigation into the differences needed to implement it are still TBD.
+
+##### Self register
+
+We have two options for supporting the Self register in the calling convention:
+
+1. Use an attribute like `[SwiftSelf]` on a parameter to specify it should go in the self register.
+2. Specify that combining the `Swift` calling convention with the `MemberFunction` calling convention means that the first argument is the `self` argument.
+
+The first option seems like a natural fit, but there is one significant limitation: Attributes cannot be used in function pointers. Function pointers are a vital scenario for us to support as we want to support virtual method tables as they are used in scenarios like protocol witness tables. The second option is a natural fit as the `MemberFunction` calling convention combined with the various C-based calling conventions specifies that there is a `this` argument as the first argument. Defining `Swift` + `MemberFunction` to imply/require the `self` argument is a perfect conceptual extension of the existing model.
```

Comment:
>there is a `this` argument as the first argument.

Is this **always** the case in Swift? I would like to ensure we don't make odd constructs just because. @stephen-hawley Do you know?

Reply:
We can make this moot pretty easily by noting that the stable witness table layouts are fragile, and undefined. We should never make use of them directly. However, with `enable-library-evolution` all virtual calls and protocol calls are available via thunks that use the `self` pointer. In terms "is the self pointer always first" that depends. When BTfS has to wrap a swift function to make it callable, the general approach is to:
1.  insert a pointer to the instance before the first argument if there is an instance pointer
2. insert a pointer to memory allocated for the return value before the first argument (which may already be self)


So, no, it is not always the case. However, I have not looked into detail how the swift ABI handles larger value types as return values in `enable-swift-evolution`.

Reply:
I think relying on `MemberFunction` might be a bit too much magic then. I think being more declarative via something akin to `SwiftThis<T>` might be more appropriate.

Reply:
> We can make this moot pretty easily by noting that the stable witness table layouts are fragile, and undefined. We should never make use of them directly. However, with `enable-library-evolution` all virtual calls and protocol calls are available via thunks that use the `self` pointer. In terms "is the self pointer always first" that depends. When BTfS has to wrap a swift function to make it callable, the general approach is to:
> 
> 1. insert a pointer to the instance before the first argument if there is an instance pointer
> 2. insert a pointer to memory allocated for the return value before the first argument (which may already be self)
> 
> So, no, it is not always the case. However, I have not looked into detail how the swift ABI handles larger value types as return values in `enable-swift-evolution`.

Re item 1: Is that argument that was introduced a replacement for the `self` argument as the functions emitted by BTfS were static and did not have a `self` argument?

Re item 2: For large return types that are passed by value, that is an implementation detail that our JIT/AOT compilers should handle automatically (the primary use case of `MemberFunction` initially was to help ensure this determination was correct in some MSVC corner cases). The user should never need to declare the "return buffer" argument manually.